docs/source/format/CDeviceDataInterface.rst

```diff
@@ -650,6 +652,331 @@ The stream source is not assumed to be thread-safe. Consumers wanting to
+    ``ArrowAsyncProducer`` object.
+
+.. c:member:: void (*ArrowAsyncDeviceStreamHandler.release)(struct ArrowAsyncDeviceStreamHandler*)
+
+    *Mandatory.* A pointer to a consumer-provided release callback for the handler.
+
+    It is valid for this to be called by a producer with or without a preceding call to
+    :c:member:`ArrowAsyncProducer.request`. This must not call any methods of an ``ArrowAsyncProducer``
+    object.
+
+.. c:member:: void* ArrowAsyncDeviceStreamHandler.private_data
+
+    *Optional.* An opaque pointer to consumer-provided private data.
+
+    Producers *MUST NOT* process this member. Lifetime of this member is handled by
+    the consumer, and especially by the release callback.
+
+The ArrowAsyncTask structure
+''''''''''''''''''''''''''''
+
+The purpose of using a Task object rather than passing the array directly to the ``on_next``
+callback is to allow for more complex and efficient thread handling. Utilizing a Task
+object allows for a producer to separate the "decoding" logic from the I/O, enabling a
+consumer to avoid transferring data between CPU cores (e.g. from one L1/L2 cache to another).
+
+This producer-provided structure has the following fields:
+
+.. c:member:: int (*ArrowArrayTask.get_data)(struct ArrowArrayTask*, struct ArrowDeviceArray*)
+
+  *Mandatory.* A callback to populate the provided ``ArrowDeviceArray`` with the available data.
+  The order of ArrowAsyncTasks provided by the producer enables a consumer to know the order of
+  the data to process.
+
+  If a non-zero value is returned from this, it should be followed only by the producer calling
+  the ``on_error`` callback of the ``ArrowAsyncDeviceStreamHandler``. Because calling this method
+  is likely to be separate from the current control flow, returning a non-zero value to signal
+  an error occuring allows the current thread to decide handle the case accordingly, while still
+  allowing all error logging and handling to be centralized in the
+  :c:member:`ArrowAsyncDeviceStreamHandler.on_error` callback.
+
+  Rather than having a separate release callback, any required cleanup should be performed as part
+  of the invocation of this callback, with ownership of the Array being passed to the consumer-provided
+  parameter calling this, and must be released separately.
+
+  It is only valid to call this method exactly once.
+
+.. c:member:: void* ArrowArrayTask.private_data
+
+  *Optional.* An opaque pointer to consumer-provided private data.
+
+  Consumers *MUST NOT* process this member. Lifetime of this member is handled by
+  the producer who created this object, and should be cleaned up if necessary during
+  the call to :c:member:`ArrowArrayTask.get_data`.
+
+The ArrowAsyncProducer structure
+''''''''''''''''''''''''''''''''
+
+This producer-provided and managed object has the following fields:
+
+.. c:member:: void (*ArrowAsyncProducer.request)(struct ArrowAsyncProducer*, uint64_t)
```

Comment:
Please let's make it `int64_t`, not `uint64_t`.

Reply:
any particular reason? The purpose of using `uint64_t` here was so that implementers don't need to manually validate that the value is > 0

Reply:
For consistency with our usual conventions. Also because AFAIU Java doesn't have unsigned integers.

Reply:
That's mostly false, as has been stated before; modern Java may not have an explicit unsigned type but has methods to manipulate integers as if they were unsigned.

Reply:
Of course it's possible to do so (at worse you can write the required code by hand), but it's more tedious than using existing facilities around signed integers, e.g. https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/foreign/ValueLayout.html#field-summary

Reply:
Overall the point still stands:
1) there is no real advantage in using an unsigned integer
2) it breaks our conventions and will make the codebase more error-prone
3) it can make things slightly harder for Java developers

so let's please switch to `int64_t`.

Reply:
The most convincing argument to me here is that it breaks the conventions, so i'll switch it to `int64_t` and update the docs that it should error if n < 0